### DISCUSSION TITLE
Can the FirehoseSubscribeReposClient run faster?

### DISCUSSION THREAD

**Discussion by emilyhunt** [MarshalX/atproto#375] [Ideas]
Since the recent wave of Bluesky signups, the firehose of posts from the network is topping out at around 1300 ops/second. This is a lot of commits! And a lot of infra is already struggling to keep up. I wanted to start a discussion about how we could future-proof firehose consumer code in case the network gets even larger.

As I understand it, the maximum amount of ops/second that the `FirehoseSubscribeReposClient` can keep up with is about 6000, after which it becomes CPU bound based on the benchmark from [way back in v0.0.26](https://github.com/MarshalX/atproto/releases/tag/v0.0.26) when `libipld` was added. Today, I moved [the astronomy feeds](https://github.com/bluesky-astronomy) to new hosting, and after some benchmarking, I can get a peak of 5100 ops/second in processing speed (which is close to that original benchmark.) Again, it seems the `FirehoseSubscribeReposClient` gets CPU bound around that point.

However, this begs the question of what happens if Bluesky continues to grow and gets ~10x larger. Currently, the library can only manage about a ~4x-greater size increase, which isn't so much. I know the Bluesky devs have talked about plans for things like filtered firehoses, but that doesn't help if someone is running a feed that uses all types of record.

**Thought 1:** I wonder whether or not the decoding steps of `FirehoseSubscribeReposClient` could be multithreaded, as I believe that's the main bottleneck right now.

**Thought 2:** Python's base multiprocessing library looks like it's also a bit of a limiting factor. When running at peak speed (5000 ops/second), about 20% of the compute time of the firehose subscription process in my feed is spent just sending things to the worker process with a `multiprocessing.Pipe` (line 63 in [this file](https://github.com/bluesky-astronomy/astronomy-feeds/blob/main/src/astrofeed_firehose/firehose_client.py#L63))*, due to Python's lack of a way to handle shared memory nicely. If the FirehoseSubscribeReposClient was faster, then this would become an even larger limit. Is there a way to multithread firehose code in a way that avoids copies? For instance, maybe if `FirehoseSubscribeReposClient` was multithreaded, maybe each thread would call `on_message_handler` itself, instead of putting it all into a single Pipe/Queue at the end and causing a different bottleneck instead.

Thanks for your help as ever, @MarshalX!

* N.B.: my implementation is a bit different from the example one in this repo as AWS is a pain and `multiprocessing.Queue` doesn't work on it. However, I think `multiprocessing.Queue` still uses `multiprocessing.Pipe`s internally to communicate with workers, so I (think?) it will have the same bottleneck as my single-Pipe solution has.

**Comment by MarshalX** [comment]
IMHO `libipld` is not a bottleneck here. Since the first version `libipld` received a bunch of performance improvements. So comparing to v0.0.26 (where the first version of `libipld` has been integrated) is not relevant to the last version. You can find benchmark results in comparison with `libipld` versions here: https://github.com/MarshalX/python-libipld/releases.

It is a shame to say but ~6k per second is my local cap. But it goes neither from `pydantic` or `libipld`! Today I updated the https://github.com/MarshalX/atproto/blob/main/examples/firehose/process_commits.py example. And I use it as a real benchmarking tool. You can start it locally by simply replacing `start_cursor=None` with `0`, and comment the 72th line to remove printing noise about new posts. 

To exclude `pydantic` and `libipld` from the benchmark I put `continue` right after getting a message from the queue

```python
    while True:
        message = pool_queue.get()
        continue
```

And the cap is still ~6k! I did not investigate it deeper but invite you to start from here :) 

My points to check:
- my network speed is shit (I don't think so)
- I do real shit inside `FirehoseSubscribeReposClient` while decoding frames (possible, but hard to believe XD)
- websockets library does brrrr and we need to replace it with something faster. I mean it uses threads, but Python ones... so if we want to speed up sync we should choose C/Rust websockets library that will not be blocked by GIL. 
- bsky server limit backfilling speed

P.S. We should definitely try `AsyncFirehoseSubscribeReposClient` since it should be truly asynced
P.P.S. To benchmark `pydantic` + `libipld` processing we could save events to disk and run benchmarking without websockets and network depend at all. To see the real performance decoding capabilities 

UPD. I found something looking good to be a candidate for the new WebSocket client https://github.com/cirospaciari/socketify.py (I hope it has clients)
UPD 2. Nope no client in `socketify` ;( new candidate: https://github.com/tarasko/picows

**Reply by emilyhunt** [reply]
I'm running the astronomy feeds now on a server with four dedicated CPU cores and a ~10 Gbps download speed (in principle) and it seems to hit a CPU bottleneck before it hits an internet bottleneck. I'll try the AsyncFirehoseSubscribeReposClient out when I can. Unfortunately, my home internet is really slow, so it's difficult for me to use nice debugging and profiling tools (although I guess I could still look at which lines of code are more or less resource intensive.)

Will try and migrate to AsyncFirehoseSubscribeReposClient to start with and see if that makes much difference!

**Reply by MarshalX** [reply]
I predict that your CPU bottleneck is in the main process where the websocket client is working. Not on processing nodes with decoding

**Reply by emilyhunt** [reply]
I can get 7000 ops/second with the `AsyncFirehoseSubscribeReposClient`, after which it still looks like a CPU botteneck - so a little bit better. I'll try and do some profiling next to work out where the bottleneck is.

**Reply by MarshalX** [reply]
@emilyhunt better. We should upgrade to websockets 13.0+. They completely rewritten the async implementation. IDK about breaking changes. atproto SDK does not allow 13.0 yet

**Reply by MarshalX** [reply]
Ah, I've completely forgotten that frame decoding also depends on dag-cbor and pydnatic. But BTW:
1. https://github.com/MarshalX/atproto/pull/376
2. https://github.com/MarshalX/atproto/pull/315 should provide even more performance

**Reply by MarshalX** [reply]
even with commented frame decoding i got the same amount of raw messages per sec :( it is just `.recv()` right now

<img width="1108" alt="image" src="https://github.com/user-attachments/assets/2ff152dc-c118-469f-8878-47d8dceb0759">

**Reply by emilyhunt** [reply]
I've tried to do some profiling of `sub_repos_async.py` from `examples/firehose`, using the latest version of atproto and Python 3.12. It was a bit of a pain to get working & understand but I think a detailed profile confirms that completely.

Here's a speedscope-compatible format file from sampling for 60 seconds on my home PC (Ryzen 5600X, 50Mbps download speed) with [austin](https://github.com/P403n1x87/austin): [samples-async-pid.json](https://github.com/user-attachments/files/16830179/samples-async-pid.json)
... which can be viewed by dropping it into [speedscope.app](https://www.speedscope.app/). I recommend the 'left heavy' view.

The profile is quite cryptic, as most time seems to be spent in `EpollSelector.select` (whatever that is.) I _think_ this is just asyncio waiting on the next commit to be downloaded over my slow internet 😅

The most interesting stuff is to the right of that. The most intensive things seem to be SSL related, with SSLObject.read taking the most time of any child function. The second most intensive thing is `_get_message_frame_from_bytes_or_raise` doing decoding. Then, there's some websocket stuff too. But yeah, reading the SSL buffer seems to be the biggest potential thing to optimize right now, so the fact that commenting out the decoding doesn't do much makes sense!

Here's an annotated screenshot of the interesting bits of the speedscope:

![annotated](https://github.com/user-attachments/assets/d7386f17-85d9-4184-9d6b-315c9ec204a7)

If those things can't be optimized, I wonder if a multithreaded websocket connection would work better, if that's even possible? (might have to jump to Rust to not have to deal with the GIL)

**Reply by MarshalX** [reply]
@emilyhunt awesome! Thank you so much for it! I will investigate it deeper a bit later. I dropped new sdk version btw. Could you pls upgrade to it? And make sure that websockets is also upgraded to 13.0.1. Also could you please try uvloop with async client? https://github.com/MagicStack/uvloop

**Reply by emilyhunt** [reply]
So, adding uvloop and using websockets 13.1 seems to help. I also worked out how to do profiling on the server with fast internet speed that the astronomy feeds are now on! 

I can get a peak processing speed with `sub_repos_async.py` of 25,000 ops/second (!!!!) on the astronomy feeds server (i.e. when doing nothing with those messages). It took about ~10 minutes to reach that speed for some reason.

Unfortunately, the Astronomy feeds implementation is still peaking at around 7000 ops/second as it's limited very strongly now by `multiprocessing.Pipe`'s time to pickle something. I believe it isn't possible to run in the background or on another thread either, I don't _think_ you can release the GIL in any way with pipes annoyingly.

So the peak speed is really high! But the GIL stops it from being useful right now for an actual feed, so we might need a different solution for how worker processes are offloaded to to get those kinds of speeds on a feed.

These are good numbers, though - if Bluesky ever reaches Twitter sizes, which averages ~5000 posts / second, then the current speeds aren't far off at least being able to keep up with a very high rate of new posts (if we get a filtered firehose from the bluesky devs with just posts) - and for a hosting cost on Hetzner of $20-$30 a month, which isn't that bad.

I'll mention this post on [the Bluesky GitHub discussion about dev infrastructure](https://github.com/bluesky-social/atproto/discussions/2754#discussioncomment-10470747) too, I figure these speed tests are useful to others =)

**Detailed profiling**
- Astronomy feeds, firehose client process - mostly taken up by multiprocessing 🫠 [samples-astrofeed_updated.json](https://github.com/user-attachments/files/16835098/samples-astrofeed_updated.json)
- `sub_repos_async.py` wall time - there's a lot of waiting for something the profiler isn't capturing, I guess it's web related still [samples-sub_repos_async_wall.json](https://github.com/user-attachments/files/16835125/samples-sub_repos_async_wall.json)
- `sub_repos_async.py` CPU time only [samples-sub_repos_async_cpu.json](https://github.com/user-attachments/files/16835137/samples-sub_repos_async_cpu.json)

again, all in [speedscope](https://www.speedscope.app/) format (download them and then upload to speedscope to view.)

**Reply by MarshalX** [reply]
Good news!

Do you need threading/multiprocessing at all? I understand that decoding is blocking operation but because it is swift and optimized I want you to try remote all multiprocessing stuff and decode CAR right in the message handler. I expect that pipe and pickle could be slower than blocking async. At least we want to know the answer.

If you are still in love with multiprocessing try https://github.com/alex-petrenko/faster-fifo 

Also could you please try https://github.com/MarshalX/atproto/blob/main/examples/firehose/process_commits_async.py with uvloop to compare perf with your feeds. pls do it on your server

**Reply by MarshalX** [reply]
@emilyhunt do we have 25k ops with **disabled frame decoding** in the firehose client? pls say that it is enabled ;((

upd. i see that it is enable in all speedcore files

**Reply by emilyhunt** [reply]
[This is the script I ran](https://github.com/bluesky-astronomy/astronomy-feeds/blob/main/scripts/sub_repos_async.py) - it's almost the same as `sub_repos_async.py` in the examples, but without printing the message in `on_message_handler` and with a counter to calculate ops/second instead (which in practice should have almost 0 overhead.)

Update: believe it should be enabled, yeah. Why would it be bad otherwise? =o

**Reply by MarshalX** [reply]
awesome! yeah if you did not modify the internals of atproto SDK it is enabled. it would be bad because it would do nothing except .recv(). which is rly nothing to benchmark expect websocket impl and bandwidth

**Reply by MarshalX** [reply]
The recipe is pretty simple to scale up to 25k:
- run async firehose client
- save messages to the local queue
- flush queue every N seconds/messages to Redis/Kafka
- scale as hard as you want using the count of workers that consume Redis/Kafka

The worker consumes raw messages, parses to commit, decodes CAR of blocks, decides to add to the feed or not, and updates some DB which stores the feed`s posts. You can have as many workers as you want. Even on the different servers.

**Reply by emilyhunt** [reply]
This sounds good! I'll have a go at some of the optimization strategies you mentioned at some point in the next few days - thanks for all the help so far!

**Reply by emilyhunt** [reply]
Gosh, it took me a while to make any progress on this (also haven't seen Bluesky scale quite so badly, so it isn't really a priority.)

I recently [upgraded the astronomy feeds firehose client](https://github.com/bluesky-astronomy/astronomy-feeds) to a) be better written, making it easier to modify and b) use [faster_fifo](https://github.com/alex-petrenko/faster-fifo). It's now about ~30% faster and has a max processing speed of ~9300 ops/second. The main bottleneck is still inter-process synchronization. 

I'll update this discussion further if I try a) other serialization strategies (might be able to find something quicker) or b) if we go the nuclear route and switch to Redis/Kafka. For now though, the astronomy feeds implementation has ~10x headroom over the current Bluesky activity peaks, so should be fine.
